sw/device/silicon_creator/lib/ownership/owner_block.c

```diff
@@ -0,0 +1,175 @@
+};
+
+rom_error_t owner_block_parse(const owner_block_t *block,
+                              owner_config_t *config,
+                              owner_application_keyring_t *keyring) {
+  if (block->header.tag != kTlvTagOwner)
+    return kErrorOwnershipInvalidTag;
+  if (block->header.length != sizeof(owner_block_t))
+    return kErrorOwnershipInvalidTagLength;
+
+  config->unknown_tags = 0;
+  config->sram_exec = block->sram_exec_mode;
+  config->flash = NULL;
+  config->info = NULL;
+  config->rescue = NULL;
+  uint32_t remain = sizeof(block->data);
+  uint32_t offset = 0;
+  while (remain) {
+    const tlv_header_t *item = (const tlv_header_t *)(block->data + offset);
+    if (item->tag == 0 || item->length == 0) {
+      break;
+    }
+    if (item->length > remain) {
+      return kErrorOwnershipInvalidTagLength;
+    }
+    remain -= item->length;
+    offset += item->length;
+    switch (item->tag) {
+      case kTlvTagApplicationKey:
+        if (keyring->length < ARRAYSIZE(keyring->key)) {
+          keyring->key[keyring->length++] =
+              (const owner_application_key_t *)item;
+        }
+        break;
+      case kTlvTagFlashConfig:
+        if (config->flash)
+          return kErrorOwnershipDuplicateItem;
+        config->flash = (const owner_flash_config_t *)item;
+        break;
+      case kTlvTagInfoConfig:
+        if (config->info)
+          return kErrorOwnershipDuplicateItem;
+        config->info = (const owner_flash_info_config_t *)item;
+        break;
+      case kTlvTagRescueConfig:
+        if (config->rescue)
+          return kErrorOwnershipDuplicateItem;
+        config->rescue = (const owner_rescue_config_t *)item;
+        break;
+      default:
+        /* unrecognized item. skip. */
+        config->unknown_tags += 1;
+    }
+  }
+  return kErrorOk;
+}
+
+#define MUBI(ex_) (ex_) ? kMultiBitBool4True : kMultiBitBool4False
+
+rom_error_t owner_block_flash_apply(const owner_flash_config_t *flash,
+                                    uint32_t config_side,
+                                    uint32_t primary_side) {
+  if (!flash)
+    return kErrorOk;
+  uint32_t start = config_side == kBootDataSlotA   ? 0
+                   : config_side == kBootDataSlotB ? kFlashBankSize
+                                                   : 0xFFFFFFFF;
+  uint32_t end = config_side == kBootDataSlotA   ? kFlashBankSize
+                 : config_side == kBootDataSlotB ? 2 * kFlashBankSize
+                                                 : 0;
+  size_t len = (flash->header.length - sizeof(owner_flash_config_t)) /
+               sizeof(owner_flash_region_t);
+  if (len >= 8) {
+    return kErrorOwnershipFlashConfigLenth;
+  }
+
+  const owner_flash_region_t *config = flash->config;
+  for (size_t i = 0; i < len; ++i, ++config) {
+    if (config->start >= start && config->start + config->size <= end) {
+      flash_ctrl_cfg_t cfg = {
+          .scrambling = MUBI(config->properties & kOwnerFlashPropertyScramble),
```

Comment:
If the hardware supports MUBI, maybe we should consider using MUBI in the data structure as well. This way we don't have to worry about faults here. 

Alternatively, we could argue if leaving as-is would be beneficial for an attacker. The read path should be protected here. In which case, the macro here just makes the attack exposure at the dual lockstep side. 

Out of all these settings, I think the most critical ones are the read/write access.

Reply:
I've migrated the flash config fields to use MUBI4 and to xor each of the configuration words with `0x11111111 * index` to create a data-dependency between the index of the item and the configuration data.